python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -1693,49 +1694,67 @@ def create_run(
         )
 
+        # Figure out the set of assets that were materialized and checks that successfully executed
         logs = self.all_logs(
             run_id=run_id,
             of_type={
-                DagsterEventType.ASSET_MATERIALIZATION_PLANNED,
                 DagsterEventType.ASSET_MATERIALIZATION,
-                DagsterEventType.ASSET_CHECK_EVALUATION_PLANNED,
                 DagsterEventType.ASSET_CHECK_EVALUATION,
             },
         )
-        planned_asset_keys: set[AssetKey] = set()
-        executed_asset_keys: set[AssetKey] = set()
-        planned_asset_check_keys: set[AssetCheckKey] = set()
-        executed_asset_check_keys: set[AssetCheckKey] = set()
+        executed_keys: set[EntityKey] = set()
+        blocking_failure_keys: set[AssetKey] = set()
         for log in logs:
             event_data = log.dagster_event.event_specific_data if log.dagster_event else None
-            if isinstance(event_data, AssetMaterializationPlannedData):
-                planned_asset_keys.add(event_data.asset_key)
-            elif isinstance(event_data, StepMaterializationData):
-                executed_asset_keys.add(event_data.materialization.asset_key)
-            elif isinstance(event_data, AssetCheckEvaluationPlanned):
-                planned_asset_check_keys.add(event_data.asset_check_key)
+            if isinstance(event_data, StepMaterializationData):
+                executed_keys.add(event_data.materialization.asset_key)
             elif isinstance(event_data, AssetCheckEvaluation):
-                executed_asset_check_keys.add(event_data.asset_check_key)
+                # blocking asset checks did not "successfully execute"
+                if event_data.blocking and not event_data.passed:
+                    blocking_failure_keys.add(event_data.asset_check_key.asset_key)
+                else:
+                    executed_keys.add(event_data.asset_check_key)
+        handled_keys = executed_keys - blocking_failure_keys
+
+        # Find the set of planned assets and checks
+        to_reexecute: set[EntityKey] = set()
+        for step in execution_plan_snapshot.steps:
+            planned_keys: set[EntityKey] = set()
+            required_keys: set[EntityKey] = set()
+            for output in step.outputs:
+                if output.properties is None:
+                    continue
+
+                key = output.properties.asset_key or output.properties.asset_check_key
+                if key is None:
+                    continue
+
+                planned_keys.add(key)
+                # keep track of all keys required on the step
+                if output.properties.is_required:
+                    required_keys.add(key)
+
+            skipped_keys = planned_keys - handled_keys
+            if skipped_keys:
+                # we need to include all keys that were marked as required on the step
+                to_reexecute.update(skipped_keys | required_keys)
```

Comment:
could consider some kind of property on ExecutionPlanSnapshot akin to the existing asset_selection one if that makes sense here (it may not if this is bespoke for some reason)